custom/icds_reports/reports/service_delivery_dashboard_data.py

```diff
@@ -0,0 +1,312 @@
+            base_dict['annaprasan_diwas_count'] = get_value_or_data_not_entered(row_data,
+                                                                                'annaprasan_diwas_count')
+            base_dict['suposhan_diwas_count'] = get_value_or_data_not_entered(row_data,
+                                                                              'suposhan_diwas_count')
+            base_dict['coming_of_age_count'] = get_value_or_data_not_entered(row_data,
+                                                                             'coming_of_age_count')
+            base_dict['public_health_message_count'] =\
+                get_value_or_data_not_entered(row_data, 'public_health_message_count')
+        elif step == 'sn':
+            base_dict = _get_pre_percents(base_dict, row_data, 'lunch', 'pse')
+            base_dict = _get_pre_values(base_dict, row_data, 'lunch', 'pse')
+
+        elif step == 'pse':
+            base_dict = _get_pre_percents(base_dict, row_data, 'pse', 'pse')
+            base_dict = _get_pre_values(base_dict, row_data, 'pse', 'pse')
+        return base_dict
+
+    all_row = dict()
+
+    data_length = len(data)
+    for index, row in enumerate(data):
+        base_row = base_data(row)
+        if not all_row.keys():
+            all_row = copy.deepcopy(base_row)
+        else:
+            all_row = update_total_row(all_row, base_row)
+        config['data'].append(base_data(row))
+    if data_length:
+        # setting location params to all
+        for location in location_fields:
+            all_row[location] = 'All'
+        # Calculating percentages for all row
+        if step == 'thr':
+            all_row = _get_pre_percents(all_row, all_row, 'thr', 'thr')
+        elif step == 'sn':
+            all_row = _get_pre_percents(all_row, all_row, 'lunch', 'pse')
+        elif step == 'pse':
+            all_row = _get_pre_percents(all_row, all_row, 'pse', 'pse')
+
+        sort_columns = [field + '_val' for field in count_columns]
+
+        percentage_fields = sort_columns
+        if order:
+            if order in percentage_fields:
+                config['data'].sort(
+                    key=lambda x: float(x[order][:-1] if x[order] != DATA_NOT_ENTERED else 0), reverse=reversed_order
+                )
+            else:
+                config['data'].sort(key=lambda x: x[order], reverse=reversed_order)
+        config['data'] = config['data'][start:(start + length)]
+        config['data'].insert(0, all_row)
+    config["aggregationLevel"] = location_filters['aggregation_level']
+    config["recordsTotal"] = data_count
+    config["recordsFiltered"] = data_count
+
+    return config
+
+
+def get_value_or_data_not_entered(source, field):
+    value = source.get(field)
+    if value is None:
+        return DATA_NOT_ENTERED
+    return value
+
+
+def _get_pre_percents(base_dict, row_data, service_name, eligibility):
```

Comment:
could you verify this that this is not breaking because these functions are defined later and called before.

Reply:
yes. The builds have passed on this as well